\documentclass[a4paper,12pt]{report}
%\documentclass[a4paper,10pt]{scrartcl}

\usepackage[utf8]{inputenc}

\title{Reconstruction of a macro-complex using interacting subunits}
\author{Lydia Fortea \and Juan Luis Melero}
\date{}

\pdfinfo{%
  /Title    (Reconstruction of a macro-complex using interacting subunits)
  /Author   (Lydia Fortea and Juan Luis Melero)
  /Creator  ()
  /Producer ()
  /Subject  (Structural Bioinformatics and Introduction to Python)
  /Keywords (modelling, reconstruction, macro-complex, structural alignment, structural bioinformatics)
}

\begin{document}
\maketitle
\tableofcontents{}

\chapter{Background}

The aim of the project is to reconstruct a marco-complex having only the pair interacting chains, using a standalone program created by ourselves.
The program we created is based on several bioinformatic features, including modelling, structural superimposition and sequence alignment, among others.

\section{Protein-Protein Interaction and Complexes}

An important point of the project is understanding the Protein-Protein Interaction and Complexes. In the quaternary structure of a protein, there are more than one separated chains of proteins that interacti between them.
The interaction of these chains can involve a lot of intermolecular bounds, such as hydrogen bounds, electrostatic interactions, pi stacking, cation-pi interaction, etc. This diversity of interactions makes the protein-protein interaction
very common in order to stabilized the molecule and generate a biological function.\\\\

The whole structure, where two or more chains are combined and have one or different functions, is called a complex. The formation of a complex can be made by protein-protein interaction only or nucleotides (DNA or RNA) can also be part of a complex if there is DNA-protein or RNA-protein interactions.\\\\

Focusing on the project, having the protein-protein interaction by pairs, we want to reconstruct the whole macro-complex.

\section{Structural superimposition}

We cannot assume that the protein-protein interacting pairs are well oriented in the space. Therefore, in order to give to each part the correct orientation, we do a structural superimposition. 


\section{Complex Modelling}

Modelling is the process through which having a protein sequence and one or more templates we can infer the structure of the protein. This is made by a program called \textit{Modeller}. 



\chapter{Algorithm and Program}


\section{Inputs and Outputs}

\subsection{Input files}

There are two types of input files to the program:

\begin{itemize}
 \item PDB files with two interacting chains
 \item For each PDB files of the interaction of two chains, PDB files of each chain separatedly.
\end{itemize}

Input files must be all in the directory specified in the arguments or in the current directory by default. The program will read all the PDB files, si it is necessary that all the PDB files in the working directory are the subunits of the macrocomplex and nothing else.

\subsection{Output file}

The output file will be one PDB file in which there will be the coordinates of the atoms of the macro-complex. The output file will be located in the directory specified in the arguments. By default, it creates a file called \textit{output.pdb} in the current directory. If the current directory is the same in which there are the input files, be careful if you want to exercute the program twice or more times at the same time, since it will take the output file also as input file.


\section{Modules and Packages}

\textit{Biopython} is the principal package used, as well as \textit{sys}, \textit{os}, \textit{re} and \textit{argparse}. 

\subsection{Biopython}

Biopython is the main open-source collection of tools written in Python to work with biological data. From Biopython we take the following subpackages:

\begin{itemize}
\item Bio.PDB, to work with PDB files
\item Bio.pairwise2, to align protein sequences one by one
\item Bio.SubsMat, to import Substitution Matrices to score the alignment
\end{itemize}


\subsection{sys}

Sys package is the System-specific parameters and functions. This package is used to read the arguments in the command line (sys.argv) and to have access to the three channels of communication with the computer: the \textit{standard in} (sys.stdin), the \textit{standard out} (sys.stdout) and the \textit{standard error} (sys.stderr).

\subsection{os}

Os package is the Miscellaneous operating system interfaces. With this package, the program can access synonimous commands of the shell, allowing the program to, for example, change working directory (\textit{cd} in shell, \textit{os.chdir} in python). 
This package is usefull to call command lines from the system but without consuming as much CPU.

\subsection{re}

Re package in the Regular expression operations package. It allows to work with regular expressions with python. In the program, it is used to find PDB and FASTA files, searching the extension of FASTA (.fa, .fasta) and PDB (.pdb) as regular expression at the end of the files.


\subsection{argparse}

Argparse package is the Parser for command-line options, arguments annd sub-commands. This packages allows to include the options for the user. The options included in the program are the following:

\begin{itemize}
 \item -i, --input; it takes the directory as input. By default, it takes the current directory.
 \item -o, --ouput; it takes the directory and the filename of the PDB file. By default, it creates a file called \textit{output.pdb} in the current directory.
 \item -s, --sequence; it takes the directory and the filename of the FASTA file in which there is the sequence of the macro-complex. By default, it takes de value \textit{None} and runs the program with default parameters.
 \item -vz, --visualize; if this option is active, at the end of the script it opens the output file with Chimera.
 \item -v, --verbose; if this option is active, it prints the log to the standard error. 
\end{itemize}


\section{Workflow}



%We superimpose those chains that are the same. We know what chains must be superimposed because we did a previous sequence alignment and we superimpose those chains with a percentage of identity greater than 99%.

%Once we have the parts well oriented, we use each part
\section{Restrictions and Limitations of the Program}


\chapter{How to use the program}

\section{Requirements}

\section{Options and Arguments}

In the program the following arguments are available:

\begin{itemize}
 \item -i, --input; it takes the directory as input. By default, it takes the current directory.
 \item -o, --ouput; it takes the directory and the filename of the PDB file. By default, it creates a file called \textit{output.pdb} in the current directory.
 \item -s, --sequence; it takes the directory and the filename of the FASTA file in which there is the sequence of the macro-complex. By default, it takes de value \textit{None} and runs the program with default parameters.
 \item -vz, --visualize; if this option is active, at the end of the script it opens the output file with Chimera.
 \item -v, --verbose; if this option is active, it prints the log to the standard error. 
\end{itemize}

All options are not forced to be, but it is highly recommended to use them, specially those related to the input/ouput files.


\chapter{Analysis of examples}

\section{Proved examples}

\section{Generalisation of the program}




\chapter{Discussion of the project}


\chapter{Conclusions}



\chapter{Appendix}

\section{Script}

\section{README}





\end{document}
