\documentclass[a4paper,12pt]{report}
%\documentclass[a4paper,10pt]{scrartcl}

\usepackage[utf8]{inputenc}

\title{Reconstruction of a macro-complex using interacting subunits}
\author{Lydia Fortea \and Juan Luis Melero}
\date{}

\pdfinfo{%
  /Title    (Reconstruction of a macro-complex using interacting subunits)
  /Author   (Lydia Fortea \and Juan Luis Melero)
  /Creator  ()
  /Producer ()
  /Subject  (Structural Bioinformatics \and Introduction to Python)
  /Keywords (modelling, reconstruction, macro-complex, structural alignment, structural bioinformatics)
}

\begin{document}
\maketitle
\tableofcontents{}

\chapter{Background}

The aim of the project is to reconstruct a marco-complex having only the pair interacting chains, using a standalone program created by ourselves.
The program we created is based on several bioinformatic features, including modelling, structural superimposition and sequence alignment, among others.

\section{Protein-Protein Interaction and Complexes}

An important point of the project is understanding the Protein-Protein Interaction and Complexes. In the quaternary structure of a protein, there are more than one separated chains of proteins that interacti between them.
The interaction of these chains can involve a lot of intermolecular bounds, such as hydrogen bounds, electrostatic interactions, pi stacking, cation-pi interaction, etc. This diversity of interactions makes the protein-protein interaction
very common in order to stabilized the molecule and generate a biological function.////

The whole structure, where two or more chains are combined and have one or different functions, is called a complex. The formation of a complex can be made by protein-protein interaction only or nucleotides (DNA or RNA) can also be part of a complex if there is DNA-protein or RNA-protein interactions.

Focusing on the project, having the protein-protein interaction by pairs, we want to reconstruct the whole macro-complex.

\section{Structural superimposition}

We cannot assume that the protein-protein interacting pairs are well oriented in the space. Therefore, in order to give to each part the correct orientation, we do a structural superimposition. 


\section{Complex Modelling}

Modelling is the process through which having a protein sequence and one or more templates we can infer the structure of the protein. This is made by a program called \textit{Modeller}. 



\chapter{Algorithm and Program}


\section{Inputs and Outputs}

\section{Modules and Packages}

\textit{Biopython} is the main package used, as well as \textit{sys}. 

\subsection{Biopython}

Biopython is the main open-source collection of tools written in Python to work with biological data. From Biopython we take the following subpackages:

\begin{itemize}
 \item Bio.PDB, to work with PDB files
 \item Bio.pairwise2, to align protein sequences one by one
 \item Bio.SubsMat, to import Substitution Matrices to score the alignment
\end{itemize}

\subsection{sys}

Sys package is the System-specific parameters and functions. This package is used to read the arguments in the command line (sys.argv) and to have access to the three channels of communication with the computer: the \textit{standard in} (sys.stdin), the \textit{standard out} (sys.stdout) and the \textit{standard error} (sys.stderr).


\section{Workflow}



%We superimpose those chains that are the same. We know what chains must be superimposed because we did a previous sequence alignment and we superimpose those chains with a percentage of identity greater than 99%.

%Once we have the parts well oriented, we use each part
\section{Restrictions and Limitations of the Program}


\chapter{How to use the program}

\section{Requirements}

\section{Arguments}


\chapter{Analysis of examples}

\section{Proved examples}

\section{Generalisation of the program}




\chapter{Discussion of the project}


\chapter{Conclusions}



\chapter{Appendix}

\section{Script}

\section{README}





\end{document}
